### PR TITLE
NIFI-11985 add ConsumeElasticsearch processor; NIFI-11016 add Query Builder for Elasticsearch processors

### PR DESCRIPTION
**Pull request by ChrisSamo632**
# Summary

[NIFI-11985](https://issues.apache.org/jira/browse/NIFI-11985) add ConsumeElasticsearch processor, allowing users to track (time) series data in Elasticsearch indices/data streams

[NIFI-11016](https://issues.apache.org/jira/browse/NIFI-11016) add Query Builder for Elasticsearch processors. The new `ConsumeElasticsearch` processor needed a set of query builder-like properties for constructing the necessary Elasticsearch `query`, so extending this to all Elasticsearch processors (where appropriate) seems sensible at the same time.

- Added new `integration-tests` for the Elasticsearch processors to cover the different ways of building Elasticsearch Query DSL
- Updated reading of test data files to be more consistent with the `support/nifi-1.x` branch so that these and future Elasticsearch processor changes will continue to be possible for the NiFi 1.x maintenance branch

**Note for merge**: `JsonUtils.readString` will not be compatible from this PR for Java 8 (required on branch `support/nifi-1.x`). Instead, the merge will need to retain the `JsonUtils.readString` method already present on the support branch. Other changes here **should** be java 8 compatible

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 17

### Licensing

- ~[ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)~
- ~[ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files~

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+@CapabilityDescription("A processor that repeatedly runs a paginated query against a field using a Range query to consume new Documents from an Elasticsearch index/query. " +
+        "The processor will retrieve multiple pages of results until either no more results are available or the Pagination Keep Alive expiration is reached, " +
+        "after which the Range query will automatically update the field constraint based on the last retrieved Document value.")
+@SeeAlso({SearchElasticsearch.class, PaginatedJsonQueryElasticsearch.class})
+@DynamicProperty(
+        name = "The name of a URL query parameter to add",
+        value = "The value of the URL query parameter",
+        expressionLanguageScope = ExpressionLanguageScope.FLOWFILE_ATTRIBUTES,
+        description = "Adds the specified property name/value as a query parameter in the Elasticsearch URL used for processing. " +
+                "These parameters will override any matching parameters in the query request body")
```
Comment (ChrisSamo632) [inline nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java:76]: Update with NIFI-12063 changes from #7695 if/when accepted

**Review by gresockj — CHANGES_REQUESTED** [review]
Overall, great work, @ChrisSamo632!  I still need to run the integration tests, and one issue came up during verification, but the general feel of `ConsumeElasticsearch` is very good.  I played around with it a bunch in NiFi and it works well and provides a great deal of flexibility.  I'll submit these comments for now, and let you know if anything comes up during integration tests.

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java [hunk 2]
```diff
@@ -460,7 +462,7 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
                 stopWatch.getDuration(TimeUnit.MILLISECONDS)
         );
 
-        input = session.putAllAttributes(input, new HashMap<>() {{
+        input = session.putAllAttributes(input, new HashMap<String, String>() {{
```
[the remark this answers is not in the document]
Comment (ChrisSamo632) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecord.java:465]: The `<String, String>` is required for Java 8 on the `support/nifi-1.x` branch - while we don't necessarily need to continue pushing to that branch, these changes **should** work fine there, so I wanted to enable that if possible

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java [hunk 3]
```diff
@@ -109,19 +210,92 @@ public interface ElasticsearchRestProcessor extends VerifiableProcessor {
             .description("All flowfiles that fail due to server/cluster availability go to this relationship.")
             .build();
 
+    String DEFAULT_QUERY_JSON = "{}";
```
[the remark this answers is not in the document]
Comment (ChrisSamo632) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java:213]: `ElasticsearchRestProcessor` is an `interface`, so this is implicitly `static final` by default (and in fact, I cannot set that explicitly)

**Comment by gresockj** [comment]
Integration tests look good, so once the above comments are addressed, I think we're good to go.

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        paginatedQueryJsonParameters.setTrackingRangeValue(getTrackingRangeValueOrDefault(context));
+        return paginatedQueryJsonParameters;
+    }
+
+    @Override
+    public void addQueryClause(final Map<String, Object> query, final Map<String, String> attributes, final ProcessContext context) throws IOException {
+        final List<Map<String, Object>> filters = new ArrayList<>(10);
+
+        // only retrieve documents with values greater than the last queried value (if present)
+        final String trackingRangeValue = getTrackingRangeValueOrDefault(context);
```
[the remark this answers is not in the document]
Comment (ChrisSamo632) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java:202]: Updated to allow for `stateMap == null` when adding the Query Clause for `ConsumeElasticsearch`

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@WritesAttributes({
+        @WritesAttribute(attribute = "mime.type", description = "application/json"),
+        @WritesAttribute(attribute = "page.number", description = "The number of the page (request) in which the results were returned that are in the output flowfile"),
```
[the remark this answers is not in the document]
Comment (ChrisSamo632) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java]: Updated (for all affected processors)

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gresockj) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java:213]: Gotcha, missed that.

File: nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gresockj) [reply nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ConsumeElasticsearch.java:202]: Thanks.  Looks like there's still an issue when verifying with certain unset properties (e.g., Sort):
```
2023-10-09 16:01:21,314 WARN [Verify Processor Config Thread-1] o.a.n.controller.StandardProcessorNode ConsumeElasticsearch[id=14e04756-018b-1000-e816-b1a31e987ad9] Unable to parse Query as JSON
com.fasterxml.jackson.databind.JsonMappingException: Null key for a Map not allowed in JSON (use a converting NullKeySerializer?) (through reference chain: java.util.HashMap["sort"]->java.util.ArrayList[0]->java.util.Collections$SingletonMap["null"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:283)
	at com.fasterxml.jackson.databind.SerializerProvider.mappingException(SerializerProvider.java:1366)
	at com.fasterxml.jackson.databind.SerializerProvider.reportMappingProblem(SerializerProvider.java:1262)
	at com.fasterxml.jackson.databind.ser.impl.FailingSerializer.serialize(FailingSerializer.java:32)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:791)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:119)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:79)
	at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serialize(IndexedListSerializer.java:18)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
	at com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:479)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:318)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4719)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:3964)
	at org.apache.nifi.processors.elasticsearch.ElasticsearchRestProcessor.getQuery(ElasticsearchRestProcessor.java:258)
	at org.apache.nifi.processors.elasticsearch.ElasticsearchRestProcessor.verifyAfterIndex(ElasticsearchRestProcessor.java:376)
```

**Review by gresockj — APPROVED** [review]
Thanks for the latest changes -- verification looks good now!

**Pull request closed without merging**